lib/agent.js

```diff
@@ -0,0 +1,80 @@
+'use strict'
+
+const { InvalidArgumentError, InvalidReturnValueError, NotSupportedError } = require('./core/errors')
+const { Pool } = require('../')
+const { weakCache, nodeMajorVersionIsGreaterThanOrEqualTo } = require('./core/util')
+
+const kAgentFactory = Symbol('agent factory')
+
+class Agent {
+  constructor (opts) {
+    /* istanbul ignore next */
+    if (!nodeMajorVersionIsGreaterThanOrEqualTo(14)) {
+      throw new NotSupportedError('Node.js version must be 14 or above')
+    }
+    this[kAgentFactory] = weakCache(origin => new Pool(origin, opts))
+  }
+
+  get (origin) {
+    if (typeof origin !== 'string' || origin === '') {
+      throw new InvalidArgumentError('Origin must be a non-empty string.')
+    }
+
+    return this[kAgentFactory](origin)
+  }
+}
+
+let globalAgent = new Agent()
+
+function setGlobalAgent (agent) {
+  /* istanbul ignore next */
+  if (!nodeMajorVersionIsGreaterThanOrEqualTo(14)) {
+    throw new NotSupportedError('Node.js version must be 14 or above')
+  }
+
+  if (typeof agent.get === 'function') {
+    globalAgent = agent
+  } else {
+    throw new InvalidArgumentError('Argument agent must implement Agent')
+  }
+}
+
+function getPath (url) {
+  return url.path || `${url.pathname || ''}${url.search || ''}`
```

Comment:
On second thought, I think `pathname` should be `||` with `'/'`. If `url` is an instance of  `URL` the right side will never be reached (`pathname` always defaults to `'/'`), otherwise we should reproduce the same behavior (i.e., the user passes a plain object)

Reply:
We should also include `hash`:

```bash
URL {
  href: 'http://localhost/?test=1#readme',
  origin: 'http://localhost',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'localhost',
  hostname: 'localhost',
  port: '',
  pathname: '/',
  search: '?test=1',
  searchParams: URLSearchParams { 'test' => '1' },
  hash: '#readme'
}
```

Reply:
Is there some existing utility method for this kind of pathing? I feel weird adding this support when it feels like theres lots of edge cases

Reply:
I don't know, unfortunately. I think this

```js
`${url.pathname || ''}${url.search || ''}${url.hash || ''}`
```

given how `hash` and `search` are parsed:

```bash
> new URL('http://localhost/?test=1#readme')
URL {
  href: 'http://localhost/?test=1#readme',
  origin: 'http://localhost',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'localhost',
  hostname: 'localhost',
  port: '',
  pathname: '/',
  search: '?test=1',
  searchParams: URLSearchParams { 'test' => '1' },
  hash: '#readme'
}
```
```bash
> new URL('http://localhost/#readme?test=1')
URL {
  href: 'http://localhost/#readme?test=1',
  origin: 'http://localhost',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'localhost',
  hostname: 'localhost',
  port: '',
  pathname: '/',
  search: '',
  searchParams: URLSearchParams {},
  hash: '#readme?test=1'
}
```

Do you see other corner cases?

Reply:
@ronag @mcollina thoughts on this?